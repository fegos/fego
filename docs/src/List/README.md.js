webpackJsonp([10],{468:function(t,s){t.exports={content:["article",["h1","\u52a0\u8f7d Loading"],["h3","\u7ec4\u4ef6\u63cf\u8ff0"],["ul",["li",["p","\u4ee5\u5217\u8868\u5f62\u5f0f\u663e\u793a\u5f53\u524d\u7684\u5185\u5bb9\u3001\u6570\u7ec4\u7b49"]],["li",["p","\u4e00\u822c\u7531\u4e3b\u8981\u4fe1\u606f\u3001\u4e3b\u8981\u64cd\u4f5c\u3001\u6b21\u8981\u4fe1\u606f\u3001\u6b21\u8981\u64cd\u4f5c\u7ec4\u6210"]],["li",["p","\u4e3b\u8981\u4fe1\u606f\u548c\u4e3b\u8981\u64cd\u4f5c\u653e\u5728\u5217\u8868\u7684\u5de6\u8fb9\uff0c\u6b21\u8981\u4fe1\u606f\u548c\u6b21\u8981\u64cd\u4f5c\u653e\u5728\u5217\u8868\u7684\u53f3\u8fb9"]],["li",["p","\u8be5\u7ec4\u4ef6\u53ea\u9002\u7528\u4e8e\u79fb\u52a8\u7aef"]]],["h3","\u793a\u4f8b\u4ee3\u7801"],["pre",{lang:"html",highlighted:'<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>List</span><span class="token punctuation">></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>List.Item</span> <span class="token attr-name">title</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>list item 1<span class="token punctuation">"</span></span> <span class="token attr-name">hasRightArrow</span><span class="token attr-value"><span class="token punctuation">=</span>{false}</span> <span class="token punctuation">/></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>List.Item</span> <span class="token attr-name">title</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>list item 2<span class="token punctuation">"</span></span> <span class="token attr-name">hasRightArrow</span> <span class="token attr-name">extra</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>extra msg<span class="token punctuation">"</span></span> <span class="token punctuation">/></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>List.Item</span> <span class="token attr-name">title</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>list item 3: show icon and ellipsis<span class="token punctuation">"</span></span> <span class="token attr-name">wrap</span><span class="token attr-value"><span class="token punctuation">=</span>{false}</span> <span class="token attr-name">iconName</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>question-o<span class="token punctuation">"</span></span> <span class="token punctuation">/></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>List.Item</span> <span class="token attr-name">title</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>list item 4: with subTitle<span class="token punctuation">"</span></span> <span class="token attr-name">subTitle</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>this is subtitle<span class="token punctuation">"</span></span> <span class="token punctuation">/></span></span>\n  <span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>List.Item</span> <span class="token attr-name">title</span><span class="token attr-value"><span class="token punctuation">=</span><span class="token punctuation">"</span>list item 5: \u5de6\u6ed1\u6709\u60ca\u559c<span class="token punctuation">"</span></span> <span class="token attr-name">slip</span> <span class="token punctuation">/></span></span>\n<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>List</span><span class="token punctuation">></span></span>'},["code",'<List>\n  <List.Item title="list item 1" hasRightArrow={false} />\n  <List.Item title="list item 2" hasRightArrow extra="extra msg" />\n  <List.Item title="list item 3: show icon and ellipsis" wrap={false} iconName="question-o" />\n  <List.Item title="list item 4: with subTitle" subTitle="this is subtitle" />\n  <List.Item title="list item 5: \u5de6\u6ed1\u6709\u60ca\u559c" slip />\n</List>']],["h2","API"],["h3","List.Item props\u5217\u8868"],["table",["thead",["tr",["th","\u5c5e\u6027"],["th","\u8bf4\u660e"],["th","\u7c7b\u578b"],["th","\u9ed8\u8ba4\u503c"]]],["tbody",["tr",["td","title"],["td","\u6807\u9898"],["td","react node"],["td"]],["tr",["td","subTitle"],["td","\u526f\u6807\u9898"],["td","react node"],["td"]],["tr",["td","children"],["td","\u5b50\u5143\u7d20\uff0c\u5c06\u76f4\u63a5\u4f5c\u4e3a\u6807\u9898\u6765\u6e32\u67d3\uff0c\u56e0\u6b64\u4f7f\u7528 children \u65f6 title \u548c subTitle \u5c06\u5931\u6548"],["td","react node"],["td"]],["tr",["td","iconName"],["td","\u6807\u9898\u5de6\u4fa7\u663e\u793a\u7684\u56fe\u6807\u540d\u79f0"],["td","string"],["td"]],["tr",["td","hasRightArrow"],["td","\u662f\u5426\u663e\u793a\u53f3\u8fb9\u7684\u7bad\u5934"],["td","boolean"],["td","true"]],["tr",["td","extra"],["td","\u663e\u793a\u5728\u53f3\u8fb9\u7684\u5185\u5bb9"],["td","react node"],["td"]],["tr",["td","wrap"],["td","\u5185\u5bb9\u8fc7\u591a\u65f6\u662f\u5426\u6298\u884c\u663e\u793a,\u6298\u884c\u7684\u8bdd\u5c06\u8fdb\u884c\u622a\u65ad\u5e76\u663e\u793a..."],["td","boolean"],["td","true"]],["tr",["td","slip"],["td","\u662f\u5426\u63d0\u4f9b\u5de6\u6ed1\u663e\u793a'\u5220\u9664'\u529f\u80fd"],["td","boolean"],["td","false"]],["tr",["td","highlight"],["td","\u70b9\u51fb\u65f6\u662f\u5426\u6709\u9ad8\u4eae\u80cc\u666f"],["td","boolean"],["td","true"]],["tr",["td","onClick"],["td","\u70b9\u51fb\u56de\u8c03"],["td","function(e){}"],["td"]],["tr",["td","onSlipItemClick"],["td","slip \u4e3a true \u65f6\uff0c\u70b9\u51fb\u5220\u9664\u7684\u56de\u8c03"],["td","function(e){}"],["td"]]]]],meta:{title:"List",subTitle:"\u6e05\u5355",filename:"src/List/README.md"},toc:["ul",["li",["a",{className:"bisheng-toc-h1",href:"#\u52a0\u8f7d-Loading",title:"\u52a0\u8f7d Loading"},"\u52a0\u8f7d Loading"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u7ec4\u4ef6\u63cf\u8ff0",title:"\u7ec4\u4ef6\u63cf\u8ff0"},"\u7ec4\u4ef6\u63cf\u8ff0"]],["li",["a",{className:"bisheng-toc-h3",href:"#\u793a\u4f8b\u4ee3\u7801",title:"\u793a\u4f8b\u4ee3\u7801"},"\u793a\u4f8b\u4ee3\u7801"]],["li",["a",{className:"bisheng-toc-h2",href:"#API",title:"API"},"API"]],["li",["a",{className:"bisheng-toc-h3",href:"#List.Item-props\u5217\u8868",title:"List.Item props\u5217\u8868"},"List.Item props\u5217\u8868"]]]}}});