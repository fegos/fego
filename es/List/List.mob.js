'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = List;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ListItem = require('./ListItem.mob');

var _ListItem2 = _interopRequireDefault(_ListItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function List(props) {
  var _props$prefixCls = props.prefixCls,
      prefixCls = _props$prefixCls === undefined ? 'ns-list' : _props$prefixCls,
      className = props.className,
      style = props.style;


  return _react2.default.createElement(
    'div',
    { className: (0, _classnames2.default)(prefixCls, className), style: style },
    props.children
  );
}

List.Item = _ListItem2.default;