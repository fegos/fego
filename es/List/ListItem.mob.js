'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactRouterDom = require('react-router-dom');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Icon = require('../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ListItem = function (_Component) {
  _inherits(ListItem, _Component);

  // 左滑有显示的内容的宽度
  function ListItem(props) {
    _classCallCheck(this, ListItem);

    var _this = _possibleConstructorReturn(this, _Component.call(this, props));

    _this.touchEvent = {};
    _this.width = 88;

    _this.onTouchStart = function (e) {
      if (_this.state.clicked) return;
      var touchEvent = e.nativeEvent;
      var touch = touchEvent.changedTouches[0];
      _this.touchEvent.pageX = touch.pageX;
      _this.touchEvent.pageY = touch.pageY;
      _this.touchEvent.screenX = touch.screenX;
      _this.touchEvent.screenY = touch.screenY;
      _this.setState({
        clicked: true
      });
    };

    _this.onTouchMove = function (e) {
      if (!_this.props.slip) return;

      // 启动左滑功能的才处理 move 事件
      var touchEvent = e.nativeEvent;
      var touch = touchEvent.changedTouches[0];
      var diffScreenY = touch.screenY - _this.touchEvent.screenY;
      if (Math.abs(diffScreenY) > touchEvent.target.offsetHeight) return;

      var diffScreenX = touch.screenX - _this.touchEvent.screenX;
      var left = diffScreenX;

      if (diffScreenX > 0) {
        // ➡️右滑，直接设为 0
        left = 0;
      } else if (diffScreenX < 0 - _this.width || _this.state.expand) {
        // ⬅️左滑，但超过了宽度，或是已展开的情况下左滑，都维持最大宽度
        left = 0 - _this.width;
      }

      _this.setState({
        moving: true,
        left: left
      });
    };

    _this.onTouchEnd = function (e) {
      var touchEvent = e.nativeEvent;
      var touch = touchEvent.changedTouches[0];
      var _state = {
        moving: false,
        clicked: false
      };
      var _this$state = _this.state,
          moving = _this$state.moving,
          expand = _this$state.expand;
      // 滑动中->滑动结束

      if (moving) {
        var diffScreenX = touch.screenX - _this.touchEvent.screenX;
        var diffScreenY = touch.screenY - _this.touchEvent.screenY;
        var left = diffScreenX;
        var expand2 = void 0;
        if (Math.abs(diffScreenY) <= touchEvent.target.offsetHeight) {
          if (diffScreenX > 0) {
            // ➡️右滑
            left = 0;
            expand2 = false;
          } else {
            // ⬅️左滑
            if (left < 0 - _this.width) {
              left = 0 - _this.width;
            }

            if (Math.abs(left) >= _this.width / 2) {
              // 滑动超过宽度的一半就展开
              left = 0 - _this.width;
              expand2 = true;
            } else {
              // 未超过宽度的一半不展开
              left = 0;
              expand2 = false;
            }
          }
        } else {
          left = 0;
          expand2 = false;
        }

        _state = _extends({}, _state, {
          expand: expand2,
          left: left
        });
      } else if (expand) {
        // 点击，但是已展开的时候，关闭展开
        _state = _extends({}, _state, {
          left: 0,
          expand: false
        });
      } else {
        // 未启动左滑,则处理点击事件
        var diffPageX = touch.pageX - _this.touchEvent.pageX;
        var diffPageY = touch.pageY - _this.touchEvent.pageY;
        var onClick = _this.props.onClick;
        // 点击回调的要求：触摸开始和结束时同一个点

        if (diffPageX === 0 && diffPageY === 0) {
          onClick(e);
        }
      }

      _this.setState(_state);
    };

    _this.onTouchCancel = function () {
      _this.setState({
        clicked: false,
        moving: false,
        left: 0
      });
    };

    _this.handleSlipTouchEnd = function () {
      var onSlipItemClick = _this.props.onSlipItemClick;


      onSlipItemClick instanceof Function && onSlipItemClick();

      _this.setState({
        moving: false,
        expand: false,
        left: 0
      });
    };

    _this.state = {
      clicked: false, // list item 是否被点击
      moving: false, // list item 是否处于 touchmove 事件中, slip 为 true 时生效
      left: 0, // slip 为 true 时 list item 的偏移量
      expand: false // touchmove 事件的结果是'删除'是否显示出来了, slip 为 true 时生效
    };
    return _this;
  }

  ListItem.prototype.render = function render() {
    var _classNames, _classNames2, _classNames3;

    var _props = this.props,
        prefixCls = _props.prefixCls,
        title = _props.title,
        subTitle = _props.subTitle,
        wrap = _props.wrap,
        hasRightArrow = _props.hasRightArrow,
        extra = _props.extra,
        verticalAlign = _props.verticalAlign,
        iconName = _props.iconName,
        slip = _props.slip,
        children = _props.children,
        className = _props.className,
        highlight = _props.highlight,
        link = _props.link,
        style = _props.style;
    // 处理有左滑时候的样式

    var finalStyle = {};
    if (slip) {
      finalStyle.left = this.state.left + 'px';
    }
    var TagName = link ? _reactRouterDom.Link : 'div';
    var extraProps = link ? { to: link } : {};
    return _react2.default.createElement(
      TagName,
      _extends({
        className: (0, _classnames2.default)(prefixCls + '-item', className, (_classNames = {}, _classNames[prefixCls + '-item-clicked'] = highlight && this.state.clicked, _classNames[prefixCls + '-item-slip'] = slip, _classNames)),
        style: style,
        onTouchStart: this.onTouchStart,
        onTouchEnd: this.onTouchEnd,
        onTouchCancel: this.onTouchCancel,
        onTouchMove: this.onTouchMove
      }, extraProps),
      _react2.default.createElement(
        'div',
        { className: prefixCls + '-item-inner', style: finalStyle },
        iconName ? _react2.default.createElement(
          'div',
          { className: prefixCls + '-icon' },
          _react2.default.createElement(_Icon2.default, { name: iconName })
        ) : null,
        children ? _react2.default.createElement(
          'div',
          { className: (0, _classnames2.default)(prefixCls + '-title', (_classNames2 = {}, _classNames2[prefixCls + '-title-nowrap'] = !wrap, _classNames2))
          },
          children
        ) : _react2.default.createElement(
          'div',
          { className: (0, _classnames2.default)(prefixCls + '-title', (_classNames3 = {}, _classNames3[prefixCls + '-title-nowrap'] = !wrap, _classNames3))
          },
          title,
          subTitle ? _react2.default.createElement(
            'div',
            { className: prefixCls + '-sub-title' },
            subTitle
          ) : null
        ),
        extra !== undefined ? _react2.default.createElement(
          'div',
          { className: prefixCls + '-extra' },
          extra
        ) : null,
        hasRightArrow ? _react2.default.createElement(
          'div',
          { className: prefixCls + '-arrow' },
          _react2.default.createElement(_Icon2.default, { name: 'right-arrow' })
        ) : null
      ),
      slip ? _react2.default.createElement(
        'div',
        { className: prefixCls + '-item-del', style: { width: this.width + 'px' }, onTouchEnd: this.handleSlipTouchEnd },
        '\u5220 \u9664'
      ) : null
    );
  };

  return ListItem;
}(_react.Component);

ListItem.defaultProps = {
  // 前缀
  prefixCls: 'ns-list',
  // 是否显示右箭头
  hasRightArrow: true,
  // 点击回调
  onClick: function onClick() {},
  // 垂直方向的对其方式
  // verticalAlign: 'center',
  // 清单内容过多时是否折行显示, true - 折行显示 false - 不折行显示，超出部分显示省略号
  wrap: true,
  // 是否能左滑
  slip: false,
  // 高亮色
  highlight: true
};
ListItem.propTypes = {
  // 前缀
  prefixCls: _propTypes2.default.oneOf(['ns-list']),
  // 标题
  title: _propTypes2.default.node,
  // 副标题
  subTitle: _propTypes2.default.node,
  // 子元素集，若使用则其他的内容相关props, eg: title, subTitle 将失效
  children: _propTypes2.default.node,
  // 右边内容
  extra: _propTypes2.default.node,
  // 左侧图标的名称
  iconName: _propTypes2.default.string,
  // 是否显示右箭头
  hasRightArrow: _propTypes2.default.bool,
  // 点击回调
  onClick: _propTypes2.default.func,
  // 垂直方向的对其方式
  /**
   * 2018.03.13 备注：该属性没看出来有什么作用，暂时先取消了
   */
  // verticalAlign: PropTypes.oneOf(['center', 'top', 'bottom']),
  // 清单内容过多时是否折行显示, true - 折行显示 false - 不折行显示，超出部分显示省略号
  wrap: _propTypes2.default.bool,
  // 是否能左滑显示'删除'
  slip: _propTypes2.default.bool,
  // 点击左滑显示'删除'的回调
  onSlipItemClick: _propTypes2.default.func,
  // 高亮色
  highlight: _propTypes2.default.bool,
  // 是否渲染成链接
  link: _propTypes2.default.string
};
exports.default = ListItem;