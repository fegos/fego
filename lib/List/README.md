---
title: List
subTitle: 清单
---

# 加载 Loading

### 组件描述
- 以列表形式显示当前的内容、数组等
- 一般由主要信息、主要操作、次要信息、次要操作组成
- 主要信息和主要操作放在列表的左边，次要信息和次要操作放在列表的右边
- 该组件只适用于移动端



### 示例代码

```html
<List>
  <List.Item title="list item 1" hasRightArrow={false} />
  <List.Item title="list item 2" hasRightArrow extra="extra msg" />
  <List.Item title="list item 3: show icon and ellipsis" wrap={false} iconName="question-o" />
  <List.Item title="list item 4: with subTitle" subTitle="this is subtitle" />
  <List.Item title="list item 5: 左滑有惊喜" slip />
</List>
```

## API

### List.Item props列表

| 属性 | 说明 | 类型 | 默认值 |
|----|-----|------|------|
| title | 标题 | react node |  |
| subTitle | 副标题 | react node |  |
| children | 子元素，将直接作为标题来渲染，因此使用 children 时 title 和 subTitle 将失效 | react node |  |
| iconName | 标题左侧显示的图标名称 | string |  |
| hasRightArrow | 是否显示右边的箭头 | boolean | true |
| extra | 显示在右边的内容 | react node |  |
| wrap | 内容过多时是否折行显示,折行的话将进行截断并显示... | boolean | true |
| slip | 是否提供左滑显示'删除'功能 | boolean | false |
| highlight | 点击时是否有高亮背景 | boolean | true |
| onClick | 点击回调 | function(e){} |  |
| onSlipItemClick | slip 为 true 时，点击删除的回调 | function(e){} |  |

