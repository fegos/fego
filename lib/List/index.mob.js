'use strict';

exports.__esModule = true;

var _List = require('./List.mob');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_List).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }