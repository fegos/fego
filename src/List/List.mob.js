import React from 'react';
import classNames from 'classnames';
import ListItem from './ListItem.mob';

export default function List(props) {
  const { prefixCls = 'ns-list', className, style } = props;

  return (
    <div className={classNames(prefixCls, className)} style={style}>
      {props.children}
    </div>
  );
}

List.Item = ListItem;
