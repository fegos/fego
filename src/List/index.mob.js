export { default } from './List.mob'