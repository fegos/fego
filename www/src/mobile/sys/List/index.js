import React, { Component } from 'react'
import { List } from 'fego'

export default class Page extends Component {
	render() {
		return (
			<div>
				<List>
					<List.Item title='list item 1' hasRightArrow={false} onClick={()=>{console.log('list item 1 click')}}></List.Item>
					<List.Item title='list item 2' hasRightArrow={true} extra={33333333333} onClick={()=>{console.log('list item 2 click')}} ></List.Item>
					<List.Item title='list item 3: nowrap and overflow will show ellipsis' wrap={false} hasRightArrow={true} extra='overflow extra content will show ellipsis' iconName='question-o' onClick={()=>{console.log('list item 3 click')}}></List.Item>
					<List.Item title='list item 4: string extra and vertical align top' hasRightArrow={true} extra='extra content' verticalAlign='top' onClick={()=>{console.log('list item 4 click')}}></List.Item>
					<List.Item title='list item 5: span extra and vertical align bottom' hasRightArrow={true} extra={<span>span extra</span>} verticalAlign='bottom' onClick={()=>{console.log('list item 5 click')}}></List.Item>
					<List.Item title='list item 6: with subTitle' subTitle='this is subtitle' hasRightArrow={true} extra='extra' onClick={()=>{console.log('list item 6 click')}} ></List.Item>
					<List.Item title='list item 7: 左滑有惊喜' slip={true} onClick={()=>{console.log('list item 7 click')}} onSlipItemClick={()=>{console.log('发起请求做一些事情')}}></List.Item>
					<List.Item title='list item 8: 使用 children' ><span>span children</span></List.Item>
				</List>
			</div>
		)
	}
}