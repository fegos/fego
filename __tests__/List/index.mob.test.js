import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import List from 'List/index.mob';

const { Item } = List;

describe('List mobile test', () => {
  it('List snapshot', () => {
    const list = Renderer.create(<List><Item title="title" /></List>);
    expect(list.toJSON()).toMatchSnapshot();
  });
});

describe('ListItem mobile test', () => {
  let setStateSpy = null;
  let setStateSpyCallCount = 0;
  beforeAll(() => {
    setStateSpy = sinon.spy(Item.prototype, 'setState');
  });
  afterAll(() => {
    setStateSpy.restore();
  });
  beforeEach(() => {
    setStateSpyCallCount = setStateSpy.callCount;
  });

  it('component should render correspond to props passed to it', () => {
    const itemAll = shallow(<Item title="title" subTitle="subtitle" iconName="yes" extra="extra" slip hasRightArrow />);
    const itemNone = shallow(<Item title="title" hasRightArrow={false} />);

    // sub title
    expect(itemAll.find('.ns-list-sub-title')).toHaveLength(1);
    expect(itemNone.find('.ns-list-sub-title')).toHaveLength(0);

    // icon
    expect(itemAll.childAt(0).childAt(0).is('.ns-list-icon')).toBe(true);
    expect(itemNone.childAt(0).childAt(0).is('.ns-list-icon')).toBe(false);

    // extra
    expect(itemAll.find('.ns-list-extra')).toHaveLength(1);
    expect(itemNone.find('.ns-list-extra')).toHaveLength(0);

    // slip
    expect(itemAll.find('.ns-list-item-del')).toHaveLength(1);
    expect(itemNone.find('.ns-list-item-del')).toHaveLength(0);

    // right arrow
    expect(itemAll.childAt(0).children().last().is('.ns-list-arrow')).toBe(true);
    expect(itemNone.childAt(0).children().last().is('.ns-list-arrow')).toBe(false);
  });

  it('children props should disable title and sub-title', () => {
    const item = shallow(<Item title="title" subTitle="subtitle">children</Item>);
    expect(item.contains('title')).toBe(false);
    expect(item.contains('subtitle')).toBe(false);
    expect(item.contains('children')).toBe(true);
  });

  it('item with link props should render as a Link', () => {
    const itemLink = shallow(<Item title="title" link="http://fe.hhtcex.com" />);
    const item = shallow(<Item title="title" />);

    expect(item.name()).toEqual('div');
    expect(itemLink.name()).toEqual('Link');
  });

  it('mock touchstart event', () => {
    const item = mount(<Item title="title" />);
    const e = {
      nativeEvent: {
        changedTouches: [{
          pageX: 0,
          pageY: 100,
          screenX: 10,
          screenY: 110,
        }],
      },
    };

    item.simulate('touchstart', e);
    // 因为 touchstart 的事件处理函数会调用 setState, 以此来间接验证事件处理函数的执行
    setStateSpyCallCount++;
    expect(setStateSpy.callCount).toBe(setStateSpyCallCount);
    expect(item.state().clicked).toBe(true);
    // this will not trigger setStateSpyCallCount increase 1
    item.simulate('touchstart', e);
    expect(setStateSpy.callCount).toBe(setStateSpyCallCount);
  });

  it('mock touchmove event', () => {
    const itemDisableMove = mount(<Item title="title" />);
    itemDisableMove.simulate('touchmove'); // will not trigger setStateSpyCallCount increase 1
    expect(setStateSpy.callCount).toBe(setStateSpyCallCount);

    const item = mount(<Item title="title" slip />);

    // 先触发 touchstart, 保存touch事件的信息
    item.simulate('touchstart', {
      nativeEvent: {
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 100,
          screenY: 100,
        }],
      },
    });
    // touchstart 会调用 setState
    expect(setStateSpy.callCount).toBe(setStateSpyCallCount + 1);
    setStateSpyCallCount++; // 手动更新值
    expect(item.state().clicked).toBe(true);
    // 先触发 touchmove
    item.simulate('touchmove', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 10,
          screenY: 80, // 将会导致 Math.abs(diffScreenY) > touchEvent.target.offsetHeight， touchmove return, 因此只有 touchstart 执行了一次setState
        }],
      },
    });
    // touchmove 不会调用 setState，所以 callcount 还是以前的值
    expect(setStateSpy.callCount).toBe(setStateSpyCallCount);

    item.simulate('touchmove', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 10, // 10 < 100, 所以模拟的是左滑
          screenY: 100,
        }],
      },
    });
    // 触发 setstate 并手动更新 callcount
    expect(setStateSpy.callCount).toBe(setStateSpyCallCount + 1);
    setStateSpyCallCount++;

    item.simulate('touchmove', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 150, // 150 > 100, 所以模拟的是右滑
          screenY: 100,
        }],
      },
    });

    expect(setStateSpy.callCount).toBe(setStateSpyCallCount + 1);
  });

  it('mock touchend event', () => {
    const onClickFunc = jest.fn();
    const item = mount(<Item title="title" onClick={onClickFunc} />);
    const e = {
      nativeEvent: {
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 100,
          screenY: 100,
        }],
      },
    };
    item.simulate('touchstart', e);
    item.simulate('touchend', e);
    expect(item.state().moving).toBe(false);
    expect(item.state().expand).toBe(false);
    expect(onClickFunc).toHaveBeenCalledTimes(1);
  });

  it('mock touchcancel event', () => {
    const item = mount(<Item title="title" />);
    item.simulate('touchstart', {
      nativeEvent: {
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 100,
          screenY: 100,
        }],
      },
    });
    expect(item.state().clicked).toBe(true);
    item.simulate('touchcancel');
    expect(setStateSpy.callCount).toBe(setStateSpyCallCount + 2);
    expect(item.state().clicked).toBe(false);
    expect(item.state().moving).toBe(false);
  });
});

describe('mock touchend event for slip item', () => {
  let item = null;
  let onClickFunc;
  beforeEach(() => {
    onClickFunc = jest.fn();
    item = mount(<Item title="title" slip onClick={onClickFunc} />);
    // 先触发 touchstart, 保存touch事件的信息
    item.simulate('touchstart', {
      nativeEvent: {
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 100,
          screenY: 100,
        }],
      },
    });
  });
  it('touchend when moving - Y 方向移动距离超过了 item 高度，作罢', () => { // 原谅夹杂了中文。。。
    // 模拟移动中触发的touchend事件
    item.setState({
      moving: true,
    });
    item.simulate('touchend', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 10,
          screenY: 80,
        }],
      },
    });
    expect(item.state().expand).toBe(false);
    expect(item.state().left).toBe(0);
    expect(onClickFunc).toHaveBeenCalledTimes(0);
  });
  it('touchend when moving - 右滑', () => {
    // 模拟移动中触发的touchend事件
    item.setState({
      moving: true,
    });
    item.simulate('touchend', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 110,
          screenY: 100,
        }],
      },
    });
    expect(item.state().expand).toBe(false);
    expect(item.state().left).toBe(0);
    expect(onClickFunc).toHaveBeenCalledTimes(0);
  });
  it('touchend when moving - 左滑超过一半展开', () => {
    // 模拟移动中触发的touchend事件
    item.setState({
      moving: true,
    });
    item.simulate('touchend', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 0,
          screenY: 100,
        }],
      },
    });
    expect(item.state().expand).toBe(true);
    expect(item.state().left).toBe(-88);
    expect(onClickFunc).toHaveBeenCalledTimes(0);
  });
  it('touchend when moving - 左滑未超过一半不展开', () => {
    // 模拟移动中触发的touchend事件
    item.setState({
      moving: true,
    });
    item.simulate('touchend', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 80,
          screenY: 100,
        }],
      },
    });
    expect(item.state().expand).toBe(false);
    expect(item.state().left).toBe(0);
    expect(onClickFunc).toHaveBeenCalledTimes(0);
  });
  it('touchend when moving - 已展开的，touchend 收起', () => {
    // 模拟已展开的touchend
    item.setState({
      expand: true,
    });
    item.simulate('touchend', {
      nativeEvent: {
        target: {
          offsetHeight: 10,
        },
        changedTouches: [{
          pageX: 0,
          pageY: 0,
          screenX: 80,
          screenY: 100,
        }],
      },
    });
    expect(item.state().expand).toBe(false);
    expect(item.state().left).toBe(0);
    expect(onClickFunc).toHaveBeenCalledTimes(0);
  });
});
