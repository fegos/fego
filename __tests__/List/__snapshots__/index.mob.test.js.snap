// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List mobile test List snapshot 1`] = `
<div
  className="ns-list"
  style={undefined}
>
  <div
    className="ns-list-item"
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={undefined}
  >
    <div
      className="ns-list-item-inner"
      style={Object {}}
    >
      <div
        className="ns-list-title"
      >
        title
      </div>
      <div
        className="ns-list-arrow"
      >
        <svg
          aria-hidden="true"
          className="nsicon"
          style={
            Object {
              "height": undefined,
              "width": undefined,
            }
          }
        >
          <use
            xlinkHref="#nsicon-right-arrow"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
